### DISCUSSION TITLE
Classes and attributes beeing kept in x-for after data change

### DISCUSSION THREAD

**Discussion by ghost** [alpinejs/alpine#1051] [1. Help] [answered]
Hi, guys.

I think i found an unexpected behavior, and would like to check it.

I have a "x-for" that shows a list of items.
I add a class manually to the first element.
If i remove all items and immediately add new items, the class is kept in the first element (the same happens for attributes).
It happens just if i remove and add the items in the same event.
I presume Alpine is doing some optimization here, reusing DOM elements, and keeping some garbage.

Here is the testcase.

```html
<html lang="en">
  <head>
    <meta charset="UTF-8">
    <script src="https://cdn.jsdelivr.net/gh/alpinejs/alpine@v2.8.0/dist/alpine.min.js"
      defer></script>
  </head>
  <body x-data="{ items: ['item1', 'item2'] }">
    <template x-for="item in items"
      x-key="item">
      <li x-text="item"></li>
    </template>
    <button @click="fn(items)">Check</button>

    <script>
      function fn(items) {
        // add a class and attribute to the first <li>
        document.querySelector('li').classList.add('test');
        document.querySelector('li').setAttribute('test', '');

        // remove all items
        items.splice(0);

        // create new items
        items.push('item3');
        items.push('item4');

        // give some time to alpine to make its job
        window.setTimeout(() => {
          // class 'test' and attribute 'test' are still there
          // both should return 'false'
          console.log(document.querySelector('li').classList.contains('test'));
          console.log(document.querySelector('li').hasAttribute('test'));
        }, 250);
      }
    </script>
  </body>
</html>
```

**Comment by ghost** [comment, marked as answer]
I tried to simplify the test case, maybe the original was too verbose.
I think this behavior can be classified as a bug.
If someone can confirm, i can fill an issue.

Thanks

**Reply by SimoTod** [reply]
There is a typo in your example, x-key doesn't exist, you need to use `:key` otherwise Alpine will just key by index (meaning that it will reuse the DOM element with the current state if they share the same index, being javascript single thread, the update won't trigger until you hit the timeout function).
It's also a bit unconventional that you pass the proxy out of the component to a variable defined in the global scope (not an issue though).

**Reply by ghost** [reply]
My bad, sorry, and thanks for your answer.

Is there a way to force Alpine to rebuild the entire DOM tree under its control?

This is my scenario:

My application marks some items adding classes manually to its elemens (ex: 'error' class).
I know i can use the 'data' to control the state and to add/remove classes through Alpine, but in my test scenarios, when i have too many records under Alpine's control, there is a noticiable delay when i change state through 'data', so i am adding classes manually to elements, specially when these classes are used just to put some visual indication on these elements.

When the user press a 'refresh' button, i would like to reload the same data again (same keys), but discarding everythig that were added manually to these elements. A full rebuild, in other words.

I tried the hack below, and it works.
I delete the items, return the control to Alpine so it can remove the DOM elements, and after an interval i re-add the data in another cycle. But it seems too hacky... Is there some alternative?

```html
items.splice(0);

window.setTimeout(() => {
  items.push('item1');
  items.push('item2');
}, 10);
```

Best regards,

**Reply by SimoTod** [reply]
Not really, what you can do is to concatenate a timestamp to the key so they are always unique and Alpine will rebuild eaxh item everytime. 

`:key="item+Date.now()"`

You need to trigger a refresh though which could be something like `items = [].concat(items)` or assigning a timestamp to a control variable

**Reply by ghost** [reply]
I could do it reloading the items in the $nextTick.

```html
    <button @click="items.splice(0);$nextTick(() => { items.push(getItems()) })">Refresh</button>
```
